modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/SelectedItemsReadOnlyObservableList.java

```diff
@@ -46,11 +46,12 @@ public SelectedItemsReadOnlyObservableList(ObservableList<Integer> selectedIndic
         this.itemsRefList = new ArrayList<>();
 
         selectedIndices.addListener((ListChangeListener<Integer>)c -> {
+            int totalRemovedSize = 0;
             beginChange();
 
             while (c.next()) {
                 if (c.wasReplaced()) {
-                    List<E> removed = getRemovedElements(c);
+                    List<E> removed = getRemovedElements(c, totalRemovedSize);
```

Comment:
hmm, don't quite understand all the fuss (but might be overlooking something, has been a while since I did a deeper dig into the selection issues ;) 

- we have a copy of the items (itemsRefs) with the same coordinates we receive from the change
- this list is in sync with the indices, that is it contains the items at the respective selectedIndices
- currently, that's done in a bulk setting at the end of the listener code

Why not sync our copy while working through the change? Doing so, would automatically collect the state of our own change, or what am I missing?

Reply:
This was my first thought, too. However, it doesn't work because `SelectedItemsReadOnlyObservableList` doesn't truthfully report its changes, so we can't simply replay its reported changes onto our copy of the items list. For example, consider the following simple piece of code:

```java
ListView<String> listView = new ListView<>(FXCollections.observableArrayList("foo", "bar"));
MultipleSelectionModel<String> model = listView.getSelectionModel();
model.setSelectionMode(SelectionMode.SINGLE);
model.getSelectedIndices().addListener((ListChangeListener<? super Integer>)System.out::println);
model.select(0);
model.select(1);
```

You'd expect the following output:
```
{ [0] added at 0 }
{ [0] replaced by [1] at 0 }
```

However, the actual output is:
```
{ [0] added at 0 }
{ [1] added at 0 }
```

While that is a bug, it almost seems like it is a bug _by design_. `MultipleSelectionModelBase` actually contains lots of code to suppress change notifications. In this particular example, a crucial change notification is suppressed in `MultipleSelectionModelBase:405`:
```
if (getSelectionMode() == SINGLE) {
    startAtomic();
    quietClearSelection();
    stopAtomic();
}
selectedIndices.set(row);
```

The reason why `MultipleSelectionModelBase` interferes with `selectedIndices` change notifications is to prevent surfacing all kinds of invalid intermediate selection states.

I tried to fix this by making `selectedIndices` changes transactional, where each of the `MultipleSelectionModel` operations would be a transaction that silently accumulates changes (clear, set, and so on), and at the end of a transaction, the changes would be distilled down into a single change notification.

However, in doing so, I ended up rewriting large parts of `MultipleSelectionModelBase`. This seemed like too big of a change. Also, I noticed that at one point you prototyped a new implementation of `SelectionModel`... maybe starting from scratch is better than adding transactions on top of the existing implementation.

Reply:
> I'm not quite sure I understand what you're getting at. Are you suggesting to apply the changes reported by `selectedIndices` to our copy of the items (`itemsRefList`) while iterating the changeset? If so, I think that doesn't work because crucial `selectedIndices` notifications are missing.
> 

exactly, that's what I meant right from the start (and obviously haven't been clear enough ;) And now I understand what you tried to explain to me, probably also right from the start .. thanks for spelling it out!

hmm .. so we are stuck with a severely misbehaving selectedItems that's trying to do its best to compensate for the misbehaviour in selectedIndices (there are not only missing but also badly incorrect, notifications see f.i. [JDK-8267781](https://bugs.openjdk.java.net/browse/JDK-8267781)). There probably should be some documentation to explain that fact. Maybe open a follow-up issue to make sure the selectedItems are revisited?

Reply:
Yes, this should be followed up on. [Here](https://bugs.openjdk.java.net/browse/JDK-8267951)'s the issue.